### PR TITLE
Enhance DBaaSConnection instanceRef handling to skip provider CR creation if instanceID is not available

### PR DESCRIPTION
**Pull request by jianrongzhang89**
Currently, DBaaS operator would create a provider instance CR, after constructing the spec of the CR regardless the validation of instanceRef data succeeds or not. If the validation fails, the operator creates a provider CR even though its spec is not valid.

This PR enhances this logic by checking the instanceRef validation result first, and creating the provider CR only when needed.

## Verification Steps
1) Create a provider account.
2) Create a DBaaSInstance CR with this account that fails during the instance provisioning, eg, for mongodb, create an instance with an existing cluster name would fail.
3) Create a DBaaSConnection CR with this instance but without instanceID populated.

Expected result:
The instance CR has the status like:
Status:
  Conditions:
    Last Transition Time:  2022-09-16T19:03:48Z
    Message:               instance ID is not available
    Reason:                DBaaSInstanceNotAvailable
    Status:                False
    Type:                  ConnectionReady

and no mongodbatlasconnection object is createad.

## Type of change

- [X ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA or in issue number have been completed
- [x] Unit tests added that prove the fix is effective or the feature works 
- [ ] Documentation added for the feature
- [X ] all relevant tests are passing
- [X] Code Review completed
- [X] Verified independently by reviewer

### CODE REVIEW

File: controllers/dbaasconnection_controller.go [hunk 1]
```diff
@@ -94,14 +94,27 @@ func (r *DBaaSConnectionReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+		if err != nil {
+			logger.Error(err, "Cannot read the instance reference")
+			cond := metav1.Condition{
+				Type:    v1alpha1.DBaaSConnectionReadyType,
+				Status:  metav1.ConditionFalse,
+				Reason:  v1alpha1.DBaaSInstanceNotAvailable,
+				Message: err.Error(),
+			}
+			r.updateConnectionStatus(ctx, &connection, &cond)
+			SetConnectionMetrics(inventory.Spec.ProviderRef.Name, inventory.Name, connection, execution)
```
Comment (redhatHameed) [inline controllers/dbaasconnection_controller.go:107]: Need to handle the metrics for `DBaaSInstanceNotAvailable` reason. https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/controllers/metrics.go#L243-L257

Comment (jianrongzhang89) [reply controllers/dbaasconnection_controller.go:107]: @redhatHameed I updated the code to support this. Note that there is a new functionality in the latest prometheus go-client library that support deletion of all metrics based on a partial label match.

File: controllers/metrics.go [hunk 2]
```diff
@@ -240,20 +240,13 @@ func SetConnectionMetrics(provider string, account string, connection dbaasv1alp
 func setConnectionStatusMetrics(provider string, account string, connection dbaasv1alpha1.DBaaSConnection) {
 	for _, cond := range connection.Status.Conditions {
 		if cond.Type == dbaasv1alpha1.DBaaSConnectionReadyType {
+			DBaaSConnectionStatusGauge.DeletePartialMatch(prometheus.Labels{metricLabelProvider: provider, metricLabelAccountName: account, metricLabelInstanceID: connection.Spec.InstanceID, metricLabelConnectionName: connection.GetName(), metricLabelNameSpace: connection.Namespace})
 			if cond.Reason == dbaasv1alpha1.Ready && cond.Status == metav1.ConditionTrue {
```
[the remark this answers is not in the document]
Comment (redhatHameed) [reply controllers/metrics.go:244]: nvm seen else statment.

**Review by redhatHameed — DISMISSED** [review]

File: controllers/metrics.go [hunk 3]
```diff
@@ -240,20 +240,14 @@ func SetConnectionMetrics(provider string, account string, connection dbaasv1alp
 func setConnectionStatusMetrics(provider string, account string, connection dbaasv1alpha1.DBaaSConnection) {
 	for _, cond := range connection.Status.Conditions {
 		if cond.Type == dbaasv1alpha1.DBaaSConnectionReadyType {
+			// First clean up existing metrics for this connection
+			DBaaSConnectionStatusGauge.DeletePartialMatch(prometheus.Labels{metricLabelConnectionName: connection.GetName(), metricLabelNameSpace: connection.Namespace})
```
Comment (redhatHameed) [inline controllers/metrics.go:244]: @jianrongzhang89 how about adding creation time stamp label ?

Comment (redhatHameed) [reply controllers/metrics.go:244]: as we have already label name, namespace of connection which will make guarantee that no other connection will be deleted.

**Review by redhatHameed — APPROVED** [review]

File: controllers/dbaasconnection_controller_test.go [hunk 4]
```diff
@@ -192,6 +192,97 @@ var _ = Describe("DBaaSConnection controller with errors", func() {
...
+		BeforeEach(assertResourceCreationIfNotExists(mongoProvider))
+		BeforeEach(assertResourceCreationIfNotExists(&defaultPolicy))
+		BeforeEach(assertDBaaSResourceStatusUpdated(&defaultPolicy, metav1.ConditionTrue, v1alpha1.Ready))
+		BeforeEach(assertResourceCreationWithProviderStatus(createdDBaaSInventory, metav1.ConditionTrue, testInventoryKind, providerInventoryStatus))
+		BeforeEach(assertResourceCreationIfNotExists(createdDBaaSInstance))
+		BeforeEach(assertResourceCreationIfNotExists(createdDBaaSConnection))
+		AfterEach(assertResourceDeletion(createdDBaaSConnection))
+		AfterEach(assertResourceDeletion(createdDBaaSInstance))
+		AfterEach(assertResourceDeletion(createdDBaaSInventory))
+		It("reconcile with error", assertDBaaSResourceStatusUpdated(createdDBaaSConnection, metav1.ConditionFalse, v1alpha1.DBaaSInstanceNotAvailable))
```
Comment (xieshenzh) [inline controllers/dbaasconnection_controller_test.go:284]: There are four error scenarios. Could you add the other three test cases?

Comment (jianrongzhang89) [reply controllers/dbaasconnection_controller_test.go:284]: OK.  I will add those cases in a separate PR (as other fixes depend on this PR).

**Review by redhatHameed — APPROVED** [review]

**Pull request merged**